[1.5 Release][RPC Reliability] RRef Idempotency and RPC Retry enablement

osalpekar opened the pull request:
Summary:
Fixes https://github.com/pytorch/pytorch/issues/32119, https://github.com/pytorch/pytorch/issues/26116, https://github.com/pytorch/pytorch/issues/33072, https://github.com/pytorch/pytorch/issues/28123

Makes RRef control messages idempotent and enables sending with retries for distributed autograd cleanup and RRef internal messages.

In order to effectively test whether these RRef and distributed autograd cleanup work with network failures/retries, I implemented an RPC Agent with a faulty send function, and enabled running tests using this as a third backend (in addition to Thrift and PGA). The tests using this backend are in a separate class (the test cases are similar but with minor changes to ensure short-running tests wait for retried RPCs to finish).

This faulty RPC Agent is pretty configurable. The tests can configure which messages types to fail, and how many messages to fail, but going forward, other RPC functionality can be overriden with faulty methods to test with failures injected.

Differential Revision: D20019236

pritamdamania87 commented on torch/csrc/distributed/autograd/context/container.cpp:
```diff
@@ -161,10 +161,13 @@ void DistAutogradContainer::sendReleaseContextRpc(int64_t context_id) {
       autograd_context_.find(context_id)->second->getKnownWorkerIds();
   auto agent = rpc::RpcAgent::getCurrentRpcAgent();
+  rpc::RpcRetryOptions options;
+  options.maxRetries = 20;
```
nit: make 20 a constexpr defined at the top of this file.

pritamdamania87 commented on torch/csrc/distributed/autograd/context/container.cpp:
```diff
@@ -161,10 +161,13 @@ void DistAutogradContainer::sendReleaseContextRpc(int64_t context_id) {
         agent->getWorkerInfo(worker_id),
-        CleanupAutogradContextReq(context_id).toMessage());
+        CleanupAutogradContextReq(context_id).toMessage(),
+        options);
```
Shouldn't we attach a callback to the Future returned here and if there is a failure, LOG(ERROR) that we couldn't release the context?

pritamdamania87 commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -157,9 +159,12 @@ void RRefContext::delUser(
         agent_->getWorkerInfo(owner),
-        RRefUserDelete(rrefId, forkId).toMessage());
+        RRefUserDelete(rrefId, forkId).toMessage(),
+        options);
```
Is this idempotent? If so, can you add a detailed comment here why it is idempotent?

pritamdamania87 commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -337,10 +350,9 @@ void RRefContext::addPendingChild(
-      iter != pendingChildren_.end(),
-      "Inconsistent states: attempt to delete a non-exist child fork.");
-  pendingChildren_.erase(iter);
+  if (iter != pendingChildren_.end()) {
```
Add comments here about why we do this (to make requests idempotent).

pritamdamania87 reviewed (changes requested):
We need unit tests for all the cases where we have added `sendWithRetries`. The unit tests should also verify that the requests that we've added retries for are idempotent requests.

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -404,25 +418,20 @@ c10::intrusive_ptr<RRef> RRefContext::delForkOfOwner(
   {
     std::lock_guard<std::mutex> lock(mutex_);
     auto rrefIter = forks_.find(rrefId);
-    TORCH_INTERNAL_ASSERT(
```
Let's add a comment to explain that we don't do membership checking here because we want to support idempotency.

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -390,11 +406,9 @@ void RRefContext::addSelfAsFork(c10::intrusive_ptr<OwnerRRef>& rref) {
 void RRefContext::addForkOfOwner(const RRefId& rrefId, const ForkId& forkId) {
```
I wonder if we should split this function into two in this case. Because this function is used in both `RREF_FORK_REQUEST` and `SCRIPT/PYTHON_REMOTE_CALL`. We should make the former idempotent but as the latter contains user code, we cannot try them twice.

osalpekar replied:
Do you mean a separate function like `addForkOfOwnerIfNotPresent` (or something along those lines) - and we call this idempotent function from when `RREF_FORK_REQUEST` is received, and use the non-idempotent function otherwise? Alternately, is it alright to make the existing function idempotent in all cases, or do you specifically want it throw an exception for the case of a repeated fork for `SCRIPT/PYTHON_REMOTE_CALL` messages?

pritamdamania87 replied:
Would be nice to have an idempotent variant only in cases where we want it to be idempotent. So we should split this into two functions and call the idempotent one only when required.

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -293,21 +301,26 @@ void RRefContext::notifyOwnerAndParentOfFork(
   }
 
+  rpc::RpcRetryOptions options;
+  options.maxRetries = 20;
```
As this appears in multiple places, I wonder if we should wrap it and the subsequent `sendWithRetries` into a utility function in `rref_context.h/cpp`

mrshenli reviewed (commented):
As @pritamdamania87 mentioned, we would need tests for this probably by using a faulty RPC agent?

osalpekar commented:
> As @pritamdamania87 mentioned, we would need tests for this probably by using a faulty RPC agent?

Yes I'm adding a faulty RPC agent as a third backend for these tests. Still working on some code cleanup, but should be up today (this version on GitHub is a bit old).

pritamdamania87 commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -415,39 +439,41 @@ void RRefContext::addSelfAsFork(c10::intrusive_ptr<OwnerRRef>& rref) {
-      if (ownerIter != owners_.end()) {
-        deletedRRef = ownerIter->second;
-        owners_.erase(ownerIter);
+    if (rrefIter != forks_.end()) {
```
Can we add a LOG(INFO) with details of the rref (rref id and fork id) in the case that it was not found? Would be goo d for debugging purposes later on.

pritamdamania87 commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -362,10 +379,13 @@ void RRefContext::addPendingChild(
+  // function are idempotent.
+  if (iter != pendingChildren_.end()) {
+    pendingChildren_.erase(iter);
+  }
```
add a LOG(INFO) here as well mentioning that the rref was not found and it is ignored assuming its a duplicate request.

pritamdamania87 commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -415,39 +439,41 @@ void RRefContext::addSelfAsFork(c10::intrusive_ptr<OwnerRRef>& rref) {
+  // (as opposed to an assertion here) ensures that messages that trigger this
+  // function are idempotent.
+  if (rrefForks.find(forkId) == rrefForks.end()) {
+    rrefForks.insert(forkId);
```
add logs indicating a duplicated request was received.

pritamdamania87 commented on torch/csrc/distributed/rpc/rref_impl.cpp:
```diff
@@ -80,6 +80,7 @@ UserRRef::UserRRef(
 
 UserRRef::~UserRRef() {
   try {
+    LOG(INFO) << "call deluser from destructor";
```
Do we need this logging? With a large number of rrefs this would cause a lot of log spew. I'd suggest making this VLOG(1) if you want to keep it.

pritamdamania87 commented on torch/testing/_internal/dist_utils.py:
```diff
@@ -128,6 +148,24 @@ def wait_until_pending_users_flushed():
+    while num_pending_users != 0:
+        time.sleep(0.1)
+        num_pending_users = int(_rref_context_get_debug_info()["num_forks"])
+        print(num_pending_users)
```
if you want to keep this print, add more information so someone seeing this on the terminal can understand what this number is.

pritamdamania87 commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1107,7 +1108,11 @@ def test_return_local_rrefs(self):
 
+        info = _rref_context_get_debug_info()
         self.assertEqual(rets, [11, 12, 13])
+        print(info["num_forks"])
```
Remove this print or add more details like print("Num RRef forks: {}".format(num_forks))

pritamdamania87 commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1107,7 +1108,11 @@ def test_return_local_rrefs(self):
         self.assertEqual(rets, [11, 12, 13])
+        print(info["num_forks"])
+        time.sleep(5)
+        # self.assertEqual(True, False)
```
What is this time.sleep for and can we remove the commented out assert?

pritamdamania87 commented on test/distributed/rpc/faulty_agent/test_faulty_agent_rpc_trap.py:
```diff
@@ -0,0 +1,31 @@
+    test_object.rpc_backend,
+    init_method=test_object.init_method,
+    num_send_recv_threads=8,
+    num_fail_sends=1,
```
Can we make this higher (like 3) to ensure better test coverage? If it increases test execution time dramatically then lets leave it as it is.

pritamdamania87 commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -2051,3 +2051,42 @@ def backward(ctx, grad):
+        success = _all_contexts_cleaned_up()
+        self.assertTrue(success)
+
+    @dist_init
```
don't we need to set faulty_messages here?

osalpekar replied:
If `faulty_messages` is not set when the faulty RPC agent is used, we fail all retryable message types (those 4 message types are listed in test_faulty_agent_rpc_trap.py and dist_utils.py). I'll add a comment here noting that all messages types will be failed.

pritamdamania87 commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1652,3 +1657,54 @@ def test_function_not_on_callee(self):
+        self.assertEqual(rets, [11, 12, 13])
+        # Allow the RREF_USER_DELETE RPCs to complete, otherwise we'll see an
+        # RRef leak.
+        time.sleep(1)
```
Can't we call `wait_until_pending_forks_flushed` here? time.sleep(1) will be flaky.

pritamdamania87 commented on torch/testing/_internal/faulty_rpc_agent/faulty_process_group_agent_pybind.cpp:
```diff
@@ -0,0 +1,80 @@
+PYBIND11_MODULE(faulty_process_group_agent_pybind, m) {
+  m.doc() = "Module for Faulty Process Group RPC Agent.";
+
+  py::module rpc_module = py::module::import("torch.distributed.rpc");
```
Can we put this under some private package like `torch.testing._internal.distributed.rpc`? This agent shouldn't be easily visible to users.

osalpekar replied:
I thought this `py::module::import` line is to import the RPC agent for use in this pybind file?

pritamdamania87 commented on torch/testing/_internal/faulty_rpc_agent/faulty_process_group_agent.h:
```diff
@@ -0,0 +1,59 @@
+  int numFailSends;
+};
+
+class FaultyProcessGroupAgent : public ProcessGroupAgent {
```
Really like this framework for tests! IIUC, this doesn't need to be just FaultProcessGroupAgent right? It can be FaultyRpcAgent and implement the RpcAgent interface. In the constructor you can take a shared_ptr<RpcAgent>. This way, this agent can be used for any backend and not just ProcessGroupAgent.

If this is too much for this PR, can you create a gh issue to refactor this later to be generic enough for any backend?

osalpekar replied:
Sounds good - would be great for testing the robustness of Tensorpipes/Thrfit Agents going forward! Here's the issue (https://github.com/pytorch/pytorch/issues/34754) - happy to work on this after the 1.5 release.

pritamdamania87 commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1652,3 +1657,54 @@ def test_function_not_on_callee(self):
+        wait_until_pending_forks_flushed()
+
+    @dist_init(faulty_messages=["RREF_CHILD_ACCEPT"])
+    def test_return_local_rrefs(self):
```
Can we have just one test instead of two and mark all the RRef messages as faulty in that test? The test could be simple and only do this:

1. A creates an RRef on B.
2. A shared the RRef to C by calling a function on C with the RRef as an argument.
3. Finally, verify all RRefs have been cleaned up by calling `wait_until_pending_forks_flushed`.

In `wait_until_pending_forks_flushed` you can extend it to wait for num_forks and num_pending_users to hit 0. In addition to this if possible we could try explicitly deleting the RRef and also ensuring `num_owner_rrefs` hits 0.

pritamdamania87 commented on torch/testing/_internal/dist_utils.py:
```diff
@@ -128,6 +148,24 @@ def wait_until_pending_users_flushed():
+    is cleaned up, and therefore is a more accurate test of whether retries
+    work in long-running programs with potential network failures.
+    '''
+    num_pending_users = int(_rref_context_get_debug_info()["num_forks"])
```
Are we mixing up num_pending_users and num_forks here? Shouldn't we check for both?

osalpekar replied:
I actually forgot to change the variables here. I added a function above this one called `wait_until_pending_users_flushed` a while back that performs the same functionality but for pending users.

pritamdamania87 reviewed (changes requested):
Looks great overall! Have some comments inline in terms of restructuring a few bits.

rohan-varma commented on torch/testing/_internal/faulty_rpc_agent/faulty_process_group_agent.cpp:
```diff
@@ -0,0 +1,77 @@
+  // we must fail the next send. If the send must be failed, we set an error
+  // on the returned future immediately and increment the counter in the map,
+  // otherwise we just call the ProcessGroupAgent send.
+  LOG(INFO) << "NUm fail sends " << failNumSends_;
```
A bunch of this code is duplicated from the internal test that has the same check to fail via a map right? Could we possibly put that code somewhere generic and re-use it?

osalpekar replied:
This class is re-used in the internal ProcessGroupAgentTest - the changes are on Phabricator.

mrshenli commented on test/distributed/rpc/faulty_agent/test_faulty_agent_rpc_trap.py:
```diff
@@ -0,0 +1,31 @@
+import sys
+
+# Buck-specific import to call into real test file
+import __test_main__
```
curious, how does this work for OSS? `__test_main__` is an internal module?

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -160,9 +166,15 @@ void RRefContext::delUser(
           agent_->getWorkerInfo(owner),
-          RRefUserDelete(rrefId, forkId).toMessage());
+          RRefUserDelete(rrefId, forkId).toMessage(),
+          options);
```
looks like the `options` code occurs in multiple places, does it make sense if we assign a default value to the `options` arg in `sendWithRetries` and remove the duplications in the call sites?

osalpekar replied:
Sounds good - we already have the default options set in RpcRetryOptions struct - we can just fall back on those.

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -86,9 +90,11 @@ std::unordered_map<std::string, std::string> RRefContext::getDebugInfo() {
   lock.unlock();
   info[kNumOwnerRRefs] = c10::to_string(ownerSize);
   info[kNumPendingUsers] = c10::to_string(numPendingUsers);
+  info[kNumForks] = c10::to_string(numForks);
```
This is actually number of OwnerRRefs with living forks?

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -505,34 +531,56 @@ void RRefContext::addForkOfOwner(const RRefId& rrefId, const ForkId& forkId) {
       }
-      forks_.erase(rrefIter);
+    } else {
+      LOG(INFO) << "Could not find OwnerRRef with RRefId = " << rrefId;
```
Let's add "likely because it was deleted by a previously retried message" to these two logs.

mrshenli commented on torch/testing/_internal/dist_utils.py:
```diff
@@ -26,12 +26,19 @@ def __init__(self, *args, **kwargs):
     'setUp' and 'tearDown' methods of unittest.
+
+    Note: pass the string representation of MessageTypes that should be used
+    with the faulty agent's send function. By default, all retriable messages
```
isn't the default value for `faulty_messages` `None`?

mrshenli replied:
I see, it's set in `test_faulty_agent_rpc_trap.py`. How does this trap work for OSS? And does it mean all our normal tests also sees the same `faulty_messages` in their `dist_util`?

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1744,3 +1745,31 @@ def test_use_rref_after_shutdown(self):
+        _delete_all_user_rrefs()
+        # wait for all the deletion messages to complete before the test
+        # calls shutdown
+        wait_until_pending_forks_flushed()
```
does this mean applications also need to do this if using retry? Any reason we don't do this in our `rpc.shutdown`?

BTW, since #31893, `_destroy_rref_context()` in `shutdown()` will wait for all `UserRRef`s to be confirmed by owner and accepted by child. Does that change have any implication here?

osalpekar replied:
Thanks for the pointer @mrshenli! After rebasing, looks like `wait_until_pending_forks_flushed` and `wait_until_pending_forks_flushed` are no longer needed. This was only because we're purposely failing the `RREF_USER_DELETE` messages the first 3 times, and we will see an RRef leak if we don't ensure those messages are processed before/during shutdown, but seems like this is taken care of now.

mrshenli commented on torch/testing/_internal/faulty_rpc_agent/faulty_process_group_agent.cpp:
```diff
@@ -0,0 +1,74 @@
+}
+
+FaultyProcessGroupAgent::FaultyProcessGroupAgent(
+  std::string workerName,
```
clang-format would indent 4 spaces here?

mrshenli commented on torch/testing/_internal/faulty_rpc_agent/faulty_process_group_agent.cpp:
```diff
@@ -0,0 +1,74 @@
+  // types. We will then check this list of types in the send function to
+  // determine whether we should fail or not.
+  for (const auto& msgString : messagesToFail) {
+    if (msgString == "RREF_FORK_REQUEST") {
```
(don' need to be addressed in this PR)
nit: we can build a const map for this? Not sure if there is a way to do automatic `Enum` <-> `std::string` conversion.

osalpekar replied:
Sounds good - I can look into this further in another PR. I tried looking for some automatic `Enum` <> `std::string` conversion (since that would be ideal here), but didn't find clean solutions.

mrshenli commented on torch/testing/_internal/faulty_rpc_agent/faulty_process_group_agent.cpp:
```diff
@@ -0,0 +1,74 @@
+  const auto key = fromVec(message.payload());
+  auto it = failMessageCountMap_.find(key);
+  if (it == failMessageCountMap_.end()) {
+    failMessageCountMap_[key] = 0;
```
do we need a lock to modify this map?

mrshenli commented on torch/testing/_internal/faulty_rpc_agent/faulty_process_group_agent_pybind.cpp:
```diff
@@ -0,0 +1,80 @@
+#include <c10d/ProcessGroup.hpp>
```
BTW, where and how does this file get compiled?

mrshenli commented on torch/testing/_internal/faulty_rpc_agent/faulty_process_group_agent_pybind.cpp:
```diff
@@ -0,0 +1,80 @@
+#include <c10d/ProcessGroup.hpp>
+#include <caffe2/torch/testing/_internal/faulty_rpc_agent/faulty_process_group_agent.h>
```
why there is a `caffe2/` before `torch/`?

mrshenli commented on torch/testing/_internal/faulty_rpc_agent/rpc.py:
```diff
@@ -0,0 +1,76 @@
+import torch.distributed.distributed_c10d as dc10d
+from torch.distributed.rpc import constants as rpc_constants
+
+from caffe2.torch.testing._internal.faulty_rpc_agent import faulty_process_group_agent_pybind
```
curious, how does this work for OSS? Why does OSS have a `caffe2.torch` module?

mrshenli commented on torch/testing/_internal/faulty_rpc_agent/rpc.py:
```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env python3
```
this file should be called `backend_registry.py` to stay consistent with `rpc/backend_registry.py`?

osalpekar replied:
Sounds good - I kept the naming conventions similar to Thrift RPC Agent, but I can change it since this is OSS.

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1744,3 +1745,31 @@ def test_use_rref_after_shutdown(self):
+            # Worker0 sends an RPC to Worker2 with the RRef as an arg
+            rpc.remote(dst_worker_c, add_rref_to_value, args=(rref, torch.ones(2, 2)))
+            # remove reference to Worker0's UserRRef
+            rref = None
```
should we use assert/to_here to check if the result is as expected?

mrshenli commented on test/distributed/rpc/faulty_agent/test_rpc_spawn.py:
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env python3
```
Could you please explain how does `test_rpc_spawn.py` work with `test_faulty_agent_rpc_trap.py`? When I launch `py.test test/distributed/rpc/faulty_agent/test_rpc_spawn.py` when does the trap kick in?

osalpekar replied:
The trap actually doesn't kick in for OSS -  I think we need some sort of additional `setUp` function just for `FaultyAgentRpcTest` that sets the `TEST_CONFIG.build_rpc_backend_options` and `TEST_CONFIG.rpc_backend_name` for the faulty agent. What would be the best way of adding an additional setup step there?

The design for the trap is similar to Thrift internally, where the Buck rule for faulty agent tests defines the trap as the main module, which sets the right configs before running the tests in `FaultyAgentRpcTest`.

mrshenli replied:
> What would be the best way of adding an additional setup step there?

For long term, we might need a little overhaul of our test code structure to support additional backend types. For short term, we can have a `FaultyRpcAgentTestFixture` that inherit from `RpcAgentTestFixture` and override `rpc_backed` and `rpc_backend_options`? Then we let the `FaultAgentRpcTest` to inherit from `FaultyRpcAgentTestFixture`?

Above won't modify `TEST_CONFIG` default values, but should return the correct value for things like `self.rpc_backend` and `self.rpc_backend_options`, which should be OK as the plan for now (IIUC) is not inheriting all tests from RpcTest yet.

If we also need to modify `TEST_CONFIG`, we can do that in `faulty_agent/test_rpc_spawn.py`? More specifically, we can do that in `FaultyRpcAgentTestWithSpawn.setUp`?

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1744,3 +1745,31 @@ def test_use_rref_after_shutdown(self):
+    not torch._six.PY3,
+    "Pytorch distributed autograd package " "does not support python2",
+)
+class FaultyAgentRpcTest(RpcAgentTestFixture):
```
Does it make sense if we let this class inherit all tests from `RpcTest`? So that we will automatically get a very good coverage on the retry algorithm? Or is it already doing so and I missed it?

osalpekar replied:
We currently only have those 2 tests for the faulty RPC Agent (`test_check_failing_messages` in rpc_test.py and `test_context_cleanup_tensor_with_grad` in dist_autograd_test.py). There's an earlier comment from Pritam suggesting just the failed messages test, basically as a representative scenario using most of our retryable messages, and ensuring they still work well with retries in the RRef context. The retry mechanism itself has a couple thorough C++ tests internally in ProcessGroupAgentTest.cpp

I tried inheriting all the tests RpcTest and DistAutogradTest, since it would definitely be great for test coverage - there are 4 tests that fail with RpcTest. I haven't dug too deep into all of them yet, but we do get RRef leaks if we don't wait for the delUser messages to reach, which do take a while since we're failing the first 3 sends (so close to 5 seconds of backoff plus the timeout duration for each message). So, we would need to make changes to make them pass.

mrshenli replied:
I see. Would I be correct if I assume this PR does not introduce any additional messages at all if there is no failure? The retry thread will just stay idle and not consuming any resources?

osalpekar replied:
We are actually forcing the first 3 sends to fail for all messages that use `sendWithRetries` (the 4 messages defined in the trap), so the retry thread is active when we use the faulty agent. By additional messages, do you mean message types outside of those 4?

mrshenli replied:
Sorry for being vague. I was mainly concerned about the behavior seen by normal users, where there is no or very little failures. Say I initialized RPC with PG agent, and I run a `rpc_sync`. After this change, is there any additional overhead compared to before?

osalpekar replied:
Ah I understand - if there are no failures, the retry thread will just stay idle. There should also be no added memory usage since messages are only added to the retryMap if they fail. With little to no failures, there shouldn't be additional overhead compared to before.

mrshenli commented on torch/testing/_internal/dist_utils.py:
```diff
@@ -56,6 +64,9 @@ def new_test_method(self, *arg, **kwargs):
 
         self.worker_id = self.rank
 
+        if faulty_messages:
```
I still didn't follow how the trap works here. Looks like the trap sets `TEST_CONFIG.build_rpc_backend_options`, which is only invoked below in `self.rpc_backend_options`. How does the trap modified the value of `faulty_messages` before this line?

osalpekar replied:
Linking the reply re setting the trap above: https://github.com/pytorch/pytorch/pull/33636#discussion_r393434900

mrshenli reviewed (changes requested):
Thanks a lot for putting together the big effort! This will be useful in multiple places. 

Most of my comments are minor stuff. Among them, some might be more important or I might misunderstand the implementation:

1. I noticed we are putting cpp code into `torch/testing`, which is Python-only for now? It might worth check with the team to see if it is OK to land cpp code there, or if there are better alternatives.
2. We might need a lock to guard `failMessageCountMap_ ` because the send function can be called from different threads?
3. Is it true that we currently only have one test `test_check_failed_messages ` for faulty RPC coverage? Does it work if we inherit all tests from `RpcTest` so that we can automatically have a good coverage on it?
4. Looks like we have a `faulty_messages` arg in `dist_init`, but we didn't set it where `dist_init` is called. Instead, we set it in a trap? How does the trap help modified the arg in `dist_init`?
5. Looks like CI didn't invoke `faulty_agent/test_rpc_spawn.py` because it is not added into `test/test_determination.py`?

mrshenli commented:
Hey @ezyang @yf225 @suo, may I ask for commendations for directory structure? We have some test-only utility C++ files (e.g., `torch/testing/_internal/faulty_rpc_agent/faulty_process_group_agent_pybind.cpp`). Is there any suggestion on where we should put those files? They are currently added into `torch/testing`, but we noticed that that folder is Python only so far. `test/cpp` seems to only contain tests instead of utility files. Looks like JIT put some of the test utilities in `torch/csrc/jit/testing` (e.g., hooks_for_testing.cpp), but those code seems also included in the built libtorch package. We would like to leave the `FaultyProcessGroupAgent` out of our binary, as it is only used in tests. Is there any build mode for testing only or any suggestions on this?

mrshenli commented:
Q: Another question is how we would like to describe this feature in v1.5? Should we not mentioning it at all since we haven't fully tested it? 

BTW, could you please run the few stress tests in `rpc_test.py` to compare the latency before and after this change?  `py.test test/distributed/rpc/test_rpc_spawn.py -k "test_stress" -vs` should do it.

ezyang commented:
We have no "C++ testing library" that can be shared across multiple C++ executables. In principle this is a useful thing to do but you will have to teach cmake about it. If you make a header only library you can put it wherever you like and it won't be included in the built binaries. C++ files only go in `torch/csrc`.

mrshenli commented on torch/testing/_internal/faulty_rpc_agent/faulty_process_group_agent.cpp:
```diff
@@ -0,0 +1,78 @@
+      failNumSends_(failNumSends),
+      messageTypesToFail_(parseMessagesToFailInput(messagesToFail)) {}
+
+std::unordered_set<MessageType> FaultyProcessGroupAgent::parseMessagesToFailInput(std::vector<std::string> messagesToFail) {
```
let's break this lone line

mrshenli commented on torch/testing/_internal/faulty_rpc_agent/faulty_process_group_agent.cpp:
```diff
@@ -0,0 +1,78 @@
+    std::chrono::milliseconds rpcTimeout,
+    std::vector<std::string> messagesToFail,
+    int failNumSends)
+    : ProcessGroupAgent(workerName, pg, numSendRecvThreads, rpcTimeout),
```
move the `workerName`, pg `shared_ptr`, and `rpcTimeout`?

mrshenli commented on torch/testing/_internal/faulty_rpc_agent/faulty_process_group_agent.cpp:
```diff
@@ -0,0 +1,78 @@
+    int failNumSends)
+    : ProcessGroupAgent(workerName, pg, numSendRecvThreads, rpcTimeout),
+      failNumSends_(failNumSends),
+      messageTypesToFail_(parseMessagesToFailInput(messagesToFail)) {}
```
also move or use const reference in `parseMessagesToFailInput`?

mrshenli commented on torch/testing/_internal/faulty_rpc_agent/faulty_process_group_agent.h:
```diff
@@ -0,0 +1,65 @@
+ private:
+  // This function parses the list of strings passed in by the python tests and
+  // resolves the Message Types that must use the faulty send.
+  std::unordered_set<MessageType> parseMessagesToFailInput(std::vector<std::string> messagesToFail);
```
this can be a const function?

mrshenli commented on torch/testing/_internal/faulty_rpc_agent/faulty_process_group_agent.h:
```diff
@@ -0,0 +1,65 @@
+ protected:
+  // This function checks the messageTypesToFail_ to determine whether to use
+  // the faulty send or not.
+  virtual bool shouldFailMessage(MessageType type);
```
const function?

mrshenli commented on torch/testing/_internal/faulty_rpc_agent/faulty_process_group_agent.cpp:
```diff
@@ -0,0 +1,78 @@
+namespace distributed {
+namespace rpc {
+
+std::string fromVec(std::vector<char> vec) {
```
use const reference vector to avoid copy?

mrshenli commented on torch/testing/_internal/faulty_rpc_agent/faulty_process_group_agent.h:
```diff
@@ -0,0 +1,65 @@
+      messagesToFail(messages_to_fail),
+      numFailSends(num_fail_sends) {
+    TORCH_CHECK(numFailSends >= 0, "numFailSends should be non-negative");
+    }
```
other parts in the code base would indent in the following way:

```c++
  FaultyProcessGroupRpcBackendOptions(
      int num_send_recv_threads,
      std::chrono::milliseconds rpc_timeout,
      std::string init_method,
      std::vector<std::string> messages_to_fail,
      int num_fail_sends = 0)
      : ProcessGroupRpcBackendOptions(num_send_recv_threads, rpc_timeout, init_method),
        messagesToFail(messages_to_fail),
        numFailSends(num_fail_sends) {
    TORCH_CHECK(numFailSends >= 0, "numFailSends should be non-negative");
  }
```

mrshenli commented on torch/testing/_internal/faulty_rpc_agent/faulty_process_group_agent.cpp:
```diff
@@ -0,0 +1,78 @@
+
+std::string fromVec(std::vector<char> vec) {
+  std::string s(vec.begin(), vec.end());
+  return s;
```
`return std::string(vec.begin(), vec.end());`?

mrshenli commented on torch/testing/_internal/faulty_rpc_agent/backend_registry.py:
```diff
@@ -0,0 +1,76 @@
+import torch.distributed.distributed_c10d as dc10d
+from torch.distributed.rpc import constants as rpc_constants
+
+from caffe2.torch.testing._internal.faulty_rpc_agent import faulty_process_group_agent_pybind
```
I assume the `caffe2` prefix would disappear after we move the faulty agent code to `torch/csrc/`?

mrshenli commented on torch/testing/_internal/faulty_rpc_agent/backend_registry.py:
```diff
@@ -0,0 +1,76 @@
+    num_fail_sends,
+    **kwargs
+):
+    return faulty_process_group_agent_pybind.FaultyProcessGroupRpcBackendOptions(
```
instead of naming it `faulty_process_group_agent_pybind`, maybe we can call it `torch.distributed.rpc._testing` after moved it to `torch/csrc/`?

mrshenli commented on test/test_determination.py:
```diff
@@ -18,6 +18,7 @@ class DeterminationTest(unittest.TestCase):
         "test_torch",
         "distributed/test_distributed",
         "distributed/rpc/test_rpc_spawn",
+        "distributed/rpc/faulty_agent/test_rpc_spawn",
```
we also need to add new tests to `test/run_test.py`.

mrshenli commented on torch/csrc/distributed/autograd/context/container.cpp:
```diff
@@ -9,6 +9,7 @@ namespace autograd {
 constexpr int kAutoIncrementBits = 48;
 constexpr int64_t kAutoIncrementMask = (1LL << kAutoIncrementBits) - 1;
 constexpr int kMaxWorkerId = 65535;
+constexpr int kNumCleanupRpcRetries = 20;
```
any reason this is called `CleanupRpc` instead of `CleanupContext`?

mrshenli commented on torch/testing/_internal/dist_utils.py:
```diff
@@ -128,6 +152,23 @@ def wait_until_pending_users_flushed():
         num_pending_users = int(_rref_context_get_debug_info()["num_pending_users"])
     return
 
+def wait_until_pending_forks_flushed():
```
this function is no longer needed?

mrshenli commented on test/distributed/rpc/faulty_agent/test_dist_autograd_spawn.py:
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env python3
```
looks like this file is not added to `run_test.py`?

mrshenli commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -2051,3 +2054,45 @@ def backward(ctx, grad):
+    not torch._six.PY3,
+    "Pytorch distributed autograd package " "does not support python2",
+)
+class FaultyAgentDistAutogradTest(FaultyRpcAgentTestFixture):
```
Let's add a test to verify that the fixture has set the backend and options properly.

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1822,3 +1825,29 @@ def test_user_rrefs_confirmed_remote(self):
+    not torch._six.PY3,
+    "Pytorch distributed autograd package " "does not support python2",
+)
+class FaultyAgentRpcTest(FaultyRpcAgentTestFixture):
```
Let's add a test to verify that the fixture has set the backend and options properly.

mrshenli commented:
test failure is real:

```
ar 17 19:02:14 Running distributed/rpc/test_rpc_spawn ... [2020-03-17 19:02:14.557900]
Mar 17 19:02:15 Traceback (most recent call last):
Mar 17 19:02:15   File "distributed/rpc/test_rpc_spawn.py", line 6, in <module>
Mar 17 19:02:15     from torch.testing._internal.distributed.rpc.rpc_test import RpcTest
Mar 17 19:02:15   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/distributed/rpc/rpc_test.py", line 29, in <module>
Mar 17 19:02:15     from torch.testing._internal.distributed.rpc.faulty_rpc_agent_test_fixture import (
Mar 17 19:02:15   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/distributed/rpc/faulty_rpc_agent_test_fixture.py", line 3, in <module>
Mar 17 19:02:15     import torch.testing._internal.faulty_rpc_agent.backend_registry  # noqa
Mar 17 19:02:15   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/faulty_rpc_agent/backend_registry.py", line 8, in <module>
Mar 17 19:02:15     from caffe2.torch.testing._internal.faulty_rpc_agent import faulty_process_group_agent_pybind
Mar 17 19:02:15 ModuleNotFoundError: No module named 'caffe2.torch'
Mar 17 19:02:15 Traceback (most recent call last):
Mar 17 19:02:15   File "test/run_test.py", line 674, in <module>
Mar 17 19:02:15     main()
Mar 17 19:02:15   File "test/run_test.py", line 667, in main
Mar 17 19:02:15     raise RuntimeError(message)
Mar 17 19:02:15 RuntimeError: distributed/rpc/test_rpc_spawn failed!
```

osalpekar commented:
The C++ files for `FaultyProcessGroupAgent` and its python bindings are now in `torch/csrc/distributed/rpc/testing`. The pybind now makes the faulty agent and its backend options available in a python module called `torch.distributed.rpc._testing` as requested. Confirmed that all rpc/dist_autograd tests are passing with and without the faulty agent, as well as the ProcessGroupAgentTests internally.

mrshenli commented on torch/csrc/distributed/rpc/testing/faulty_process_group_agent.cpp:
```diff
@@ -0,0 +1,87 @@
+  std::unique_lock<std::mutex> lock(failMapMutex_);
+  auto it = failMessageCountMap_.find(key);
+  if (it == failMessageCountMap_.end()) {
+    failMessageCountMap_[key] = 0;
```
nit: If we get the the iterator again here, can we avoid the two map look ups below?

mrshenli commented on torch/csrc/distributed/rpc/testing/faulty_process_group_agent.cpp:
```diff
@@ -0,0 +1,87 @@
+}
+
+bool FaultyProcessGroupAgent::shouldFailMessage(MessageType type) const {
+  // Return true if the input message type is in the messageTypesToFail list
```
`messageTypesToFail` -> `messageTypesToFail_`?

mrshenli commented on torch/csrc/distributed/rpc/testing/faulty_process_group_agent.h:
```diff
@@ -0,0 +1,71 @@
+      : ProcessGroupRpcBackendOptions(
+            num_send_recv_threads,
+            rpc_timeout,
+            init_method),
```
move the string?

mrshenli commented on torch/csrc/distributed/rpc/testing/faulty_process_group_agent.h:
```diff
@@ -0,0 +1,71 @@
+            num_send_recv_threads,
+            rpc_timeout,
+            init_method),
+        messagesToFail(messages_to_fail),
```
move the vector?

mrshenli commented on torch/csrc/distributed/rpc/testing/init.cpp:
```diff
@@ -0,0 +1,118 @@
+#include <torch/csrc/utils/pybind.h>
+
+template <typename T>
+using shared_ptr_class_ = py::class_<T, std::shared_ptr<T>>;
```
nit: other init.cpp files put this inside the inner anonymous namespace. We can do the same to stay consistent

mrshenli commented on torch/distributed/rpc/_testing/__init__.py:
```diff
@@ -0,0 +1,18 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import numbers
+import sys
```
these two imports are not used?

mrshenli commented on torch/distributed/rpc/_testing/__init__.py:
```diff
@@ -0,0 +1,18 @@
+import sys
+
+import torch
+import torch.distributed as dist
```
and this too?

mrshenli commented on torch/distributed/rpc/_testing/faulty_agent_backend_registry.py:
```diff
@@ -0,0 +1,78 @@
+    # Initialize FaultyProcessGroup.
+    if dist.is_initialized():
+        raise RuntimeError(
+            "Faulty process group must not be initialized before init_rpc."
```
Faulty -> default?

mrshenli commented on torch/distributed/rpc/_testing/faulty_agent_backend_registry.py:
```diff
@@ -0,0 +1,78 @@
+        return FaultyProcessGroupAgent(
+            name,
+            group,
+            rpc_backend_options.num_send_recv_threads,
```
nit: (does not need to be addressed in this PR and this problem already exist prior to this PR)

It's probably terser if we allow to directly pass in the rpc_backend_options object for agents.

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1832,3 +1835,36 @@ def test_user_rrefs_confirmed_remote(self):
+
+@unittest.skipIf(
+    not torch._six.PY3,
+    "Pytorch distributed autograd package " "does not support python2",
```
why these are two separate strings?

mrshenli commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -2061,3 +2064,52 @@ def backward(ctx, grad):
+        self.assertTrue(success)
+
+    # no faulty_messages defined so this fails all retryable messages - see
+    # test/distributed/rpc/faulty_agent/test_faulty_agent_rpc_trap.py for the
```
we no longer have a trap file in OSS

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1832,3 +1835,36 @@ def test_user_rrefs_confirmed_remote(self):
+class FaultyAgentRpcTest(FaultyRpcAgentTestFixture):
+
+    # no faulty_messages defined so this fails all retryable messages - see
+    # test/distributed/rpc/faulty_agent/test_faulty_agent_rpc_trap.py for the
```
no trap file in OSS any more

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1832,3 +1835,36 @@ def test_user_rrefs_confirmed_remote(self):
+            # remove reference to Worker0's UserRRef
+            rref = None
+        # explicitly delete all User RRefs
+        _delete_all_user_rrefs()
```
graceful shutdown should have already called this?

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1832,3 +1835,36 @@ def test_user_rrefs_confirmed_remote(self):
+            # check if the output is as expected
+            self.assertEqual(rref.to_here(), torch.add(torch.ones(2, 2), torch.ones(2, 2)))
+            # remove reference to Worker0's UserRRef
+            rref = None
```
if we don't explicitly set the rref to None, do we hit leak?

mrshenli commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -2061,3 +2064,52 @@ def backward(ctx, grad):
+        # NOT require grad, we still cleanup the dist autograd contexts created
+        # on other nodes. This is because the autograd context is still
+        # communicated over RPC even if tensor arguments do not require grad, as
+        #  it is possible that the response could.
```
duplicated spaces before "it"

mrshenli reviewed (commented):
Look great to me, we are very close to landing states I think. Thanks a lot for putting this together! This is a significant amount of effort!

mrshenli commented:
All my comments today are nits, will approve when all tests pass. Thanks @osalpekar !

mrshenli commented on torch/csrc/distributed/rpc/testing/init.cpp:
```diff
@@ -0,0 +1,118 @@
+#include <c10d/ProcessGroup.hpp>
```
this header is unused?

mrshenli commented on torch/csrc/distributed/rpc/testing/init.cpp:
```diff
@@ -0,0 +1,118 @@
+#include <c10d/ProcessGroup.hpp>
+#include <pybind11/chrono.h>
+#include <pybind11/stl.h>
```
and this one too?

mrshenli commented:
Test failure is real:

```
Mar 20 03:13:31 ======================================================================
Mar 20 03:13:31 FAIL [0.001s]: test_test_file (__main__.DeterminationTest)
Mar 20 03:13:31 Test files trigger themselves and dependent tests
Mar 20 03:13:31 ----------------------------------------------------------------------
Mar 20 03:13:31 Traceback (most recent call last):
Mar 20 03:13:31   File "test_determination.py", line 65, in test_test_file
Mar 20 03:13:31     self.determined_tests(["test/test_jit.py"]), ["test_jit_simple", "test_jit"]
Mar 20 03:13:31 AssertionError: Lists differ: ['test_jit_simple', 'test_jit', 'distributed/rpc/faulty_agent/test_rpc_spawn'] != ['test_jit_simple', 'test_jit']
Mar 20 03:13:31 
Mar 20 03:13:31 First list contains 1 additional elements.
Mar 20 03:13:31 First extra element 2:
Mar 20 03:13:31 'distributed/rpc/faulty_agent/test_rpc_spawn'
Mar 20 03:13:31 
Mar 20 03:13:31 - ['test_jit_simple', 'test_jit', 'distributed/rpc/faulty_agent/test_rpc_spawn']
Mar 20 03:13:31 + ['test_jit_simple', 'test_jit']
```

mrshenli commented:
Master is green now, could you please rebase to [this commit](https://github.com/pytorch/pytorch/commit/1c958f8ef94e70efd420170642945fa9db99bf10) so that we can have clear signals from CI? Thanks!

osalpekar committed:
[1.5 Release][RPC Reliability] RRef Idempotency and RPC Retry enablement (#33636)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/33636

Fixes https://github.com/pytorch/pytorch/issues/32119, https://github.com/pytorch/pytorch/issues/26116,
https://github.com/pytorch/pytorch/issues/33072

Makes RRef control messages idempotent and enables sending with retries for distributed autograd cleanup and RRef internal messages.

In order to effectively test whether these RRef and distributed autograd cleanup work with network failures/retries, I implemented an  RPC Agent with a faulty send function, and enabled running tests using this as a third backend (in addition to Thrift and PGA). The tests using this backend are in a separate class (the test cases are similar but with minor changes to ensure short-running tests wait for retried RPCs to finish).

This faulty RPC Agent is pretty configurable. The tests can configure which messages types to fail, and how many messages to fail, but going forward, other RPC functionality can be overriden with faulty methods to test with failures injected.

Differential Revision: D20019236

fbshipit-source-id: e6cf6704da59cb28edcf2d9b11a9ace507a0f988

osalpekar commented:
@mrshenli - rebased this, and looks like CircleCI is green. Jenkins don't seem like actual test failures, but just wanted to double-check. Let me know what you think!

mrshenli reviewed (approved):
Let's land! Thanks for this big effort!
